File: GoEngineering/13_pipeline/explanation/13_pipeline_explanation.go
/*
Problem 13 — Pipeline
======================

WHAT WE'RE BUILDING

A generic multi-stage concurrent pipeline: a producer emits items onto a
channel, one or more fan-out stages transform them concurrently, and a
fan-in stage merges everything back into a single stream a caller can
drain. This is the shape behind almost every batch/stream processing job
you'll write in Go — read rows, transform each with bounded concurrency,
collect results — and it is also where Go's channel-ownership discipline
either pays off or turns into a goroutine leak.

WHY THIS MATTERS IN REAL SYSTEMS

  - Every stage of a pipeline is a goroutine (or a pool of them) connected
    by channels. The rule that keeps this leak-free is simple to state and
    easy to violate: only the goroutine that CREATES a channel is allowed
    to close it, and it must close it on every exit path (success, error,
    or ctx cancellation) — otherwise a downstream `range` over that channel
    blocks forever and its goroutine never exits.
  - Fan-out (one input channel, many worker goroutines reading it) is how
    you parallelize a slow per-item transform (a network call, a hash, a
    resize) without parallelizing the whole pipeline by hand.
  - Fan-in (many channels, one merged output) is the mirror problem: N
    workers each producing on their own channel need to be joined into one
    stream without the joiner closing the merged channel too early (while
    a worker might still send) or too late (leaking the reader).
  - Cancellation must propagate UPSTREAM, not just stop the downstream
    reader. If a caller stops reading mid-pipeline without canceling a
    shared context, every upstream stage blocks forever trying to send on
    a channel nobody drains — the single most common pipeline goroutine
    leak in real codebases.

CONCEPTS COVERED

  - Channel ownership: exactly one goroutine creates + closes each channel;
    every other goroutine only sends or only receives.
  - Fan-out with bounded worker count, fan-in via a `sync.WaitGroup` closer
    goroutine (the same "wait for every producer, then close" shape as
    12_worker_pool's Results channel, applied to N-to-1 instead of pool
    results).
  - Context-cancellation-propagates-upstream: canceling ctx must unblock
    every stage, not just the final reader.
  - Generics (Go 1.18+) so the pipeline works over any element type without
    per-type duplication.

# THE SPEC

Package `pipeline`.

	func Generate[T any](ctx context.Context, items []T) <-chan T
	    - Emits each item from items onto the returned channel, in order.
	    - Owns and closes the returned channel: closes it after emitting the
	      last item, OR immediately if ctx is canceled before finishing (no
	      goroutine leak either way).
	    - A send that would block past ctx cancellation must abort instead of
	      blocking forever — select on both the send and ctx.Done().

	type Result[T any] struct {
	    Value T
	    Err   error
	}

	func FanOut[In, Out any](ctx context.Context, in <-chan In, workers int, fn func(context.Context, In) (Out, error)) <-chan Result[Out]
	    - Spawns `workers` goroutines (minimum 1), each ranging over `in` and
	      calling fn on every item, sending a Result[Out] to the returned
	      channel.
	    - Does NOT own `in` — never closes it (the caller/Generate owns it).
	    - DOES own the returned channel: closes it only after every worker
	      goroutine has exited (a WaitGroup + one closer goroutine, exactly
	      like NewPool's results-closer in 12_worker_pool).
	    - ctx cancellation stops workers from picking up new items from `in`
	      and stops them from blocking on a send to the output channel.

	func FanIn[T any](ctx context.Context, cs ...<-chan T) <-chan T
	    - Merges zero or more input channels into one output channel, in
	      whatever order items arrive (interleaved, not round-robin).
	    - Owns and closes the returned channel: closes it once every input
	      channel has been drained AND closed, or immediately on ctx
	      cancellation.
	    - Zero input channels returns an already-closed empty channel, not a
	      channel that blocks forever.

	func Collect[T any](ctx context.Context, in <-chan Result[T]) ([]T, error)
	    - Drains `in` into a slice of values, in arrival order.
	    - On the FIRST Result with a non-nil Err, stops draining, returns
	      (nil, that error) — the caller is expected to have derived ctx from
	      a cancel func and call it so upstream stages stop promptly (see
	      HINTS); Collect itself only reads, it does not cancel.
	    - If ctx is canceled while draining (and no error has been seen yet),
	      returns (nil, ctx.Err()).
	    - Fully draining `in` to completion with no errors returns
	      (all values in arrival order, nil).

ACCEPTANCE CRITERIA

  - A full pipeline (Generate -> FanOut -> FanIn -> Collect) run to
    completion returns every input item's transformed value exactly once,
    with no duplicates and no drops.
  - Canceling ctx at any point unblocks every stage within a short bound —
    no goroutine outlives the pipeline call (verified via
    runtime.NumGoroutine() before/after with a short settle loop).
  - FanOut never has more than `workers` calls to fn executing at once.
  - FanIn never closes its output channel while an input channel might
    still send on it, and never leaks a goroutine waiting on an input
    channel that will never close.
  - Everything passes `go test -race` (or, absent a test file for this
    topic, `go vet ./...` finds nothing and the demo in a scratch `main`
    shows correct, leak-free behavior under -race manually).

HINTS

  - Generate: `select { case out <- item: case <-ctx.Done(): close(out); return }`
    inside the loop, `close(out)` again after the loop for the normal-exit
    path (closing twice is fine as long as it's the same goroutine and you
    guard with a labeled break / return, not two independent close calls on
    both paths — return immediately after the ctx.Done() close).
  - FanOut's closer goroutine: `go func() { wg.Wait(); close(out) }()` —
    same shape as 12_worker_pool's `go func() { eg.Wait(); close(p.results) }()`.
  - FanIn: one goroutine per input channel, each doing
    `for v := range c { select { case out <- v: case <-ctx.Done(): return } }`,
    all coordinated by one shared WaitGroup whose Wait() gates the single
    `close(out)` call — never let more than one goroutine call close(out).
  - Collect: a plain `for { select { case r, ok := <-in: ...; case <-ctx.Done(): ... } }`
    loop; remember `ok == false` means `in` is closed and draining is done,
    not an error.

COMMON PITFALLS

  - Closing a channel from more than one goroutine, or closing it more than
    once — both panic. Every channel in this package has exactly one
    designated closer.
  - FanOut or FanIn closing `in` — they never own an input channel, only
    the caller/upstream stage that created it does.
  - Downstream stopping early (e.g. Collect returning on first error)
    without the caller canceling ctx — upstream FanOut workers then block
    forever trying to send results nobody reads. Collect returning an error
    is a signal to the CALLER to cancel, not something Collect does for you.
  - FanIn's closer goroutine calling close(out) before every input channel
    has actually finished draining — a `sync.WaitGroup` sized to len(cs),
    with each per-input goroutine calling Done() on its own exit, is the
    only reliable way to know "every input is drained."

STRETCH GOALS

  - Add a `Tee[T any]` stage that duplicates one input channel onto two
    output channels (both must be read from, or both must respect ctx, to
    avoid one slow reader stalling the other).
  - Add per-item timeouts inside FanOut via `context.WithTimeout` layered
    on the shared ctx, independent per item.
  - Make FanIn preserve a deterministic order (round-robin across inputs)
    as an alternate function, and explain in comments why that is strictly
    slower (it must wait for the "next" input in rotation even if a later
    one is ready sooner) than the interleaved default.
  - Wire Generate -> FanOut -> FanIn -> Collect together behind a single
    `Run[In, Out any](ctx, items []In, workers int, fn func(context.Context, In) (Out, error)) ([]Out, error)`
    convenience function that also calls cancel() internally on the first
    error, demonstrating the "downstream error cancels upstream" pattern
    end-to-end in one call.
*/
package pipeline

import (
	"context"
)

// Result pairs a fan-out stage's output with any error it returned.
type Result[T any] struct {
	Value T
	Err   error
}

// Generate emits each item from items onto the returned channel, in order,
// then closes it. See the header comment for the full contract.
func Generate[T any](ctx context.Context, items []T) <-chan T {
	panic("TODO: implement Generate")
}

// FanOut spawns `workers` goroutines (minimum 1) consuming from in and
// applying fn to each item, sending a Result[Out] on the returned channel.
// It does not own (and never closes) in; it does own and close the
// returned channel, only after every worker has exited.
func FanOut[In, Out any](ctx context.Context, in <-chan In, workers int, fn func(context.Context, In) (Out, error)) <-chan Result[Out] {
	panic("TODO: implement FanOut")
}

// FanIn merges zero or more input channels into one output channel. It owns
// and closes the returned channel, only once every input channel has been
// fully drained and closed (or ctx is canceled).
func FanIn[T any](ctx context.Context, cs ...<-chan T) <-chan T {
	panic("TODO: implement FanIn")
}

// Collect drains in into a slice of values in arrival order, stopping and
// returning an error on the first Result with a non-nil Err, or on ctx
// cancellation. It does not cancel ctx itself — see the header comment.
func Collect[T any](ctx context.Context, in <-chan Result[T]) ([]T, error) {
	panic("TODO: implement Collect")
}

File: PyEngineering/13_pipeline/13_pipeline_explanation.py
"""
13 · Pipeline (fan-out / fan-in)
================================

WHAT WE'RE BUILDING
--------------------
A multi-stage async processing pipeline: a producer feeds items into stage
1, stage 1 fans out to N worker coroutines that transform each item and
feed stage 2, and stage 2 fans everything back in to a single collector.
This is the canonical "fan-out/fan-in" shape behind an ETL job, a batch
image-resize service, a log-processing pipeline, or any system that reads
records from one place, does bounded-concurrency work on each, and writes
results somewhere else.

Same shape as `../GoEngineering/13_pipeline` (channels + goroutines); here
the channel is `asyncio.Queue` and the goroutine is a `asyncio.Task` inside
an `asyncio.TaskGroup`.

WHY THIS MATTERS IN REAL SYSTEMS
---------------------------------
- **Fan-out gives you bounded concurrency for free.** A fixed pool of N
  worker tasks pulling from one queue is a natural concurrency limiter —
  you never have more than N transforms in flight, unlike
  `asyncio.gather(*(transform(x) for x in items))` which launches all of
  them at once.
- **Fan-in needs a completion signal, not a fixed count.** The collector
  doesn't know in advance how many results N workers will produce combined
  (workers finish at different times) — it needs a reliable "every worker
  is done and the queue is drained" signal. Two idiomatic options: sentinel
  values (one per worker) or `queue.join()` paired with `task_done()`. This
  problem uses sentinels because they compose cleanly with `TaskGroup`
  cancellation — `queue.join()` alone can hang forever if a worker dies
  without calling `task_done()`, whereas a sentinel-starved queue.get()
  is trivially visible and a `TaskGroup` propagates a worker's exception
  immediately, cancelling siblings instead of hanging.
- **Leak-free shutdown is the actual bar.** Every task spawned must either
  run to completion or be cancelled and awaited before the pipeline
  function returns. `asyncio.TaskGroup` (3.11+) gives you this for free:
  if any task raises, every other task in the group is cancelled and the
  group `__aexit__` awaits all of them before re-raising (as an
  `ExceptionGroup`). This is directly testable: `asyncio.all_tasks()`
  must return to baseline after the pipeline finishes or fails.
- **Backpressure is not optional.** Every `asyncio.Queue` here is bounded.
  An unbounded queue between a fast producer and slow workers is just
  deferred unbounded memory growth — the same lesson as problem 16's
  bounded subscriber queues, applied to a work queue instead of an event
  stream (contrast problem 16's drop-oldest policy, which is wrong for
  work that must be processed exactly once — here we block instead).

CONCEPTS COVERED
-----------------
- `asyncio.Queue` as a bounded, backpressured channel between stages
- Fan-out: N worker tasks reading from one queue
- Fan-in: multiple producers feeding one downstream queue/collector
- Sentinel-based "N workers done" signaling vs. `queue.join()`
- `asyncio.TaskGroup` for structured concurrency: exceptions propagate,
  siblings get cancelled, everything is awaited before the group exits
- Verifying zero leaked tasks via `asyncio.all_tasks()` before/after

THE SPEC
--------
Build a two-stage pipeline:

    async def run_pipeline(
        items: Iterable[int],
        *,
        transform: Callable[[int], Awaitable[int]],
        num_workers: int = 4,
        queue_maxsize: int = 8,
    ) -> list[int]:
        '''Run `items` through `transform` with `num_workers` concurrent
        workers (stage 1: fan-out) and collect every result into a list,
        order not guaranteed (stage 2: fan-in). Returns once every item has
        been transformed and collected, with zero leaked tasks.

        If `transform` raises for any item, the whole pipeline stops:
        remaining input stops being fed, in-flight workers finish or are
        cancelled, and the exception propagates to the caller (wrapped in
        an ExceptionGroup, since asyncio.TaskGroup raises one on any child
        failure — see problem 17 for the taxonomy of unwrapping this).
        '''

    class Pipeline:
        '''A reusable, explicitly-lifecycled version of the same pipeline,
        for callers that want to keep it running across multiple batches
        instead of a one-shot function call.'''
        def __init__(
            self,
            *,
            transform: Callable[[int], Awaitable[int]],
            num_workers: int = 4,
            queue_maxsize: int = 8,
        ) -> None: ...
        async def __aenter__(self) -> "Pipeline": ...
        async def __aexit__(self, *exc: object) -> None: ...
        async def submit(self, item: int) -> None: ...
            # blocks (backpressure) if the input queue is full
        async def results(self) -> AsyncIterator[int]:
            # async-iterate results as they complete; ends when the
            # pipeline is closed and drained
        async def close(self) -> None: ...
            # signal no more input; wait for in-flight work to finish

Behavior:
    - `run_pipeline` is the simple one-shot entry point: feed a fixed
      iterable, get a fixed list back. Internally it wires a producer task,
      `num_workers` worker tasks, and the current coroutine as collector,
      all under one `asyncio.TaskGroup`.
    - `Pipeline` is the long-lived version: an external caller `submit()`s
      items over time (e.g. from a web handler) and separately consumes
      `results()` as an async iterator, until `close()` is called and
      drains.
    - Both must guarantee: no orphaned tasks after normal completion, no
      orphaned tasks after an exception, and correct backpressure (a
      full input queue blocks `submit`/the producer, it never drops or
      grows unbounded).

ACCEPTANCE CRITERIA
--------------------
1. `run_pipeline` processes every input item exactly once and returns all
   results (order-independent), with `num_workers` transforms genuinely
   running concurrently (provable via a timing test: N items each
   sleeping `t` seconds with `num_workers` workers takes ~`(N/num_workers)*t`,
   not `N*t`).
2. A `transform` exception cancels sibling workers and the producer, and
   propagates to the caller — no swallowed errors, no hang.
3. `asyncio.all_tasks()` (minus the current task) returns to its exact
   pre-pipeline membership after both a successful run and a failed run.
4. `Pipeline.submit()` blocks once the bounded input queue is full
   (backpressure), provably via a test that a slow consumer stalls a fast
   producer rather than letting the queue grow past `queue_maxsize`.
5. `Pipeline.close()` lets in-flight items finish, then ends `results()`'s
   async iteration cleanly (no exception, no hang).
6. mypy-clean, black-formatted, ruff-clean.
"""

from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable, Iterable
from typing import Self

# A private sentinel pushed onto the results queue by each worker once it
# has permanently stopped, so the fan-in side knows when every worker is
# done without needing a pre-known item count.
# TODO: define `_DONE: Final[object] = object()`


async def run_pipeline(
    items: Iterable[int],
    *,
    transform: Callable[[int], Awaitable[int]],
    num_workers: int = 4,
    queue_maxsize: int = 8,
) -> list[int]:
    """Run `items` through `transform` with bounded fan-out concurrency."""
    # TODO:
    # 1. Create `in_q: asyncio.Queue[int | object]` bounded by queue_maxsize
    #    (or split sentinel type cleanly with a wrapper — your call).
    # 2. Inside `async with asyncio.TaskGroup() as tg:`
    #    - spawn a producer task that puts every item then, per worker,
    #      puts one sentinel (so each worker gets exactly one stop signal)
    #    - spawn num_workers worker tasks: loop `get()`, break on sentinel,
    #      else call `await transform(item)` and append to a shared list
    #      (a plain list append is fine — no other task interleaves on a
    #      single-threaded event loop between the append itself and the
    #      next `await`, so no lock is needed here)
    # 3. Return the collected results after the TaskGroup exits.
    raise NotImplementedError("TODO: implement run_pipeline")


class Pipeline:
    """Long-lived pipeline: external submit() + results() over its lifetime."""

    def __init__(
        self,
        *,
        transform: Callable[[int], Awaitable[int]],
        num_workers: int = 4,
        queue_maxsize: int = 8,
    ) -> None:
        # TODO: store config, create self._in_q (bounded by queue_maxsize —
        # this is the backpressure surface submit() blocks on) and
        # self._out_q (leave UNBOUNDED: `asyncio.Queue()`, no maxsize — a
        # bounded out_q can deadlock close() against a caller who hasn't
        # started draining results() yet), self._closed, and placeholders
        # for the worker tasks started in __aenter__.
        raise NotImplementedError("TODO: implement Pipeline.__init__")

    async def __aenter__(self) -> Self:
        # TODO: start num_workers worker tasks (store them; do NOT use a
        # TaskGroup here since __aenter__ must return before workers
        # finish — a TaskGroup's `async with` block wouldn't exit until
        # all workers do). Track tasks so __aexit__/close() can await them.
        raise NotImplementedError("TODO: implement Pipeline.__aenter__")

    async def __aexit__(self, *exc: object) -> None:
        await self.close()

    async def submit(self, item: int) -> None:
        """Enqueue one item; blocks if the bounded input queue is full."""
        # TODO: raise RuntimeError if closed; else `await self._in_q.put(item)`.
        raise NotImplementedError("TODO: implement Pipeline.submit")

    async def results(self) -> AsyncIterator[int]:
        """Async-iterate results as workers produce them."""
        # TODO: loop `await self._out_q.get()`; stop once every worker has
        # signaled done (track a countdown of live workers, decremented on
        # each `_DONE` sentinel received) — do not stop on the first
        # sentinel, since num_workers workers each emit one.
        raise NotImplementedError("TODO: implement Pipeline.results")
        yield 0  # pragma: no cover - makes this an async generator for mypy

    async def close(self) -> None:
        """Signal no more input, then wait for all in-flight work to finish."""
        # TODO: idempotent. Put one sentinel per worker into self._in_q
        # (blocking put is fine/expected here — draining the real backlog
        # first is the point), then `await asyncio.gather(*self._tasks)`.
        raise NotImplementedError("TODO: implement Pipeline.close")


# ---------------------------------------------------------------------------
# HINTS
# -----
# - `asyncio.TaskGroup` (3.11+) is the structured-concurrency primitive for
#   `run_pipeline`: it blocks at `__aexit__` until every child task is
#   done, and if any child raises, it cancels all siblings first. This is
#   exactly "leak-free on both success and failure" for free.
# - For `Pipeline` (long-lived, spans multiple `submit`/`results` calls
#   outside any single `async with tg:` block), plain `asyncio.create_task`
#   + manually tracking/awaiting the tasks in `close()` is correct — a
#   `TaskGroup` can't span an `__aenter__`/`__aexit__` pair the way you'd
#   need here, since its own `async with` body IS the task lifetime.
# - One sentinel *per worker*, not one sentinel total — if you put a single
#   sentinel into a multi-consumer queue, whichever worker happens to
#   `get()` it first exits, leaving the rest blocked on `get()` forever.
# - Use `asyncio.Queue(maxsize=...)` (never 0/unbounded) for both stages —
#   backpressure is a hard requirement, not a nice-to-have.
#
# COMMON PITFALLS
# ---------------
# - Using `asyncio.gather(*(transform(x) for x in items))` instead of a
#   worker pool — that launches unbounded concurrency, defeating the whole
#   point of "fan-out with N workers."
# - Forgetting per-worker sentinels and instead trying to signal "done" via
#   a shared counter/event without care for the exact moment every worker
#   has actually seen its own sentinel — races are easy to introduce here.
# - Calling `close()` a second time and re-sending sentinels /
#   double-awaiting already-finished tasks — must be idempotent, same
#   lesson as problem 16.
# - Appending to a shared results list from multiple worker coroutines
#   without realizing this is *fine* on a single-threaded event loop
#   (list.append doesn't await mid-operation) — don't over-engineer with a
#   `threading.Lock` that adds nothing here and would be actively wrong to
#   reach for in a coroutine (there's no OS thread contention to protect
#   against; that's problem 14's territory, on a real thread pool).
#
# STRETCH GOALS
# --------------
# - Add a third stage (fan-out -> fan-out -> fan-in) and confirm the
#   backpressure chain: a slow stage-2 stalls stage-1 workers once the
#   inter-stage queue fills, which in turn stalls the producer.
# - Add per-item timeout via `asyncio.timeout()` (3.11+) around each
#   `transform` call, routing timed-out items to a separate "failed" list
#   instead of aborting the whole pipeline.
# - Compare `queue.join()` + `task_done()` as an alternative fan-in
#   completion signal to the sentinel approach used here, and explain in
#   comments why it's a worse fit under `TaskGroup` cancellation (a
#   cancelled worker task that dies mid-item without calling `task_done()`
#   leaves `join()` waiting forever, whereas sentinel-starvation is visible
#   immediately as "the results queue.get() never returns").
